In [10]:
import numpy as np

## Part 1

In [11]:
def f(x):
    """This function calculates the following polynomial expression for a given x value: f(x) = x^3 - x^2 - 1
    
    Parameters: 
    x -- takes an integer or float value
    
    Returns: 
    f(x) -- integer or float value
    """
    return x**3 - x**2 - 1

In [12]:
# Example test (using integer input)
f(2) # expected value of 3

3

In [13]:
def df(x):
    """This function calculates the derivative of the previous f(x) function
    
    Parameters: 
    x -- takes an integer or float value
    
    Returns: 
    df/dx (x) = 3x^2 - 2x -- integer or float value
    """
    return 3*(x**2) - 2*x

In [14]:
# Example test (using float input)
df(2.0) # expected value of 8.0

8.0

## Part 2

In [15]:
def newton(f, df, x0, epsilon=1e-6, max_iter=30):
    """This function performs a newton iteration of the function f with derivative df
    
    Parameters: 
    f -- callable function that we want to find the roots of
    df -- callable, derivative of the function f
    x0 -- initial guess for function root
    epsilon -- small integer value to represent the accuracy to stop iterating at (|f(x_n)| < epsilon)
    max_iter -- integer value that represents that maximum iterations to run
    
    Returns: 
    The value of the root x_n (such that |f(x_n)|<epsilon) and the number of iterations it took to find the
    root. Will return "iteration failed" if iterations exceed the specificed maximum iterations and |f(x_n)|>epsilon
    """
    N = 0
    x_n = x0
    while N < max_iter:
        if abs(f(x_n)) < epsilon:
            print("Found root in ", N, "iterations")
            return x_n
        x_n = x_n - f(x_n)/df(x_n)
        N += 1
                   
    else:
        print("Iteration failed")
        return None   

## Part 3

In [16]:
# Test two different values of x0
newton(f, df, 2)

Found root in  4 iterations


1.4655713749070918

In [17]:
newton(f, df, 4)

Found root in  7 iterations


1.4655712318772516

In [18]:
# Test using epsilon = 1e-8
newton(f, df, 2, epsilon=1e-8)

Found root in  5 iterations


1.4655712318767877

We can see that reducing epsilon to 1e-8 still works but takes more iterations to converge, in particular for the x0=2 example it took one additional iteration to find the root of ~1.46557. This makes sense as newton iteration has quadratic convergence so we expect the iterations to go up with a higher precision (lower epsilon)